### DISCUSSION TITLE
Symmetric commands: tee

### DISCUSSION THREAD

**Discussion by Poshi** [johnkerl/miller#939] [Ideas]
`tee` is another command that works in one way but not the other. It allows you to send the data stream to a file or a **command**. But I cannot see any construct that allows us to inject the results from a command into the data stream.
The closest thing I can imagine is `system()`, but it returns the whole stdout blob with no easy way of feeding it into the data stream (you can always postprocess it, cut by new lines, extract fields and `emit` data but...).

This is just food for though. I don't need the functionality, but it triggers an alarm in me seeing this imbalance.

**Comment by johnkerl** [comment]
@Poshi agreed.

I've thought about this on & off for a year or two. You're right, `system()` pays the fork-and-exec overhead on every invocation, which doesn't scale.

In analogy with the concept of 'coroutine' in programming languages, I like the idea of a 'coprogram'. Question though about output-buffering (an arbitrary program can't be relied on to `fflush` every output line), and boundaries (do we insist the coprogram be writing 'records' one per line or not)? Etc.

Anyway I share your interest & I'd love to get some brainstorming going on how to make this happen ...
